[GlobalOpt] Handle operators separately when removing GV users

gandhi56 opened the pull request:
Refactor globalopt by eliminating redundant code. Fix https://github.com/llvm/llvm-project/issues/64680.

gandhi56 commented:
I see a flaw in my solution, I will update this PR with a more effective solution. Stay tuned!

gandhi56 commented:
Internal CI passed.

gandhi56 commented:
ping

gandhi56 commented:
@nikic @fhahn ping

gandhi56 commented:
Ping

dtcxzyw commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -201,61 +201,51 @@ CleanupPointerRootUsers(GlobalVariable *GV,
-        if (I->hasOneUse())
-          Dead.push_back(std::make_pair(I, SI));
-      }
-    } else if (MemSetInst *MSI = dyn_cast<MemSetInst>(U)) {
-      if (isa<Constant>(MSI->getValue())) {
-        Changed = true;
-        MSI->eraseFromParent();
-      } else if (Instruction *I = dyn_cast<Instruction>(MSI->getValue())) {
-        if (I->hasOneUse())
-          Dead.push_back(std::make_pair(I, MSI));
-      }
-    } else if (MemTransferInst *MTI = dyn_cast<MemTransferInst>(U)) {
-      GlobalVariable *MemSrc = dyn_cast<GlobalVariable>(MTI->getSource());
-      if (MemSrc && MemSrc->isConstant()) {
-        Changed = true;
-        MTI->eraseFromParent();
-      } else if (Instruction *I = dyn_cast<Instruction>(MTI->getSource())) {
-        if (I->hasOneUse())
-          Dead.push_back(std::make_pair(I, MTI));
-      }
```
Why did you drop the support for memset/memcpy? It causes some regressions https://github.com/dtcxzyw/llvm-opt-benchmark/pull/466#discussion_r1546056742.

Please add some tests with memset/memcpy.

gandhi56 replied:
For some reason we didn't have any tests with memset/memcpy. I will revert this snippet and add some tests.

fhahn commented on llvm/test/Transforms/GlobalOpt/cleanup-pointer-root-loads.ll:
```diff
@@ -0,0 +1,42 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py UTC_ARGS: --version 4
```
please add new test coverage in a separate PR (i.e. precommit them, so that here we only see the functional diff)

gandhi56 replied:
https://github.com/llvm/llvm-project/pull/87443

fhahn reviewed (commented):
Please clarify the commit title and description to provide more details on the change. `Refactor CleanupPointerRootUsers` makes it sound like there's no change in functionality, but IIUC this adds new functionality.

gandhi56 commented:
Apologies for so many forced pushes, I was trying to rebase this PR against the precommit PR. I suppose I will wait for the precommit to merge and then I can rebase this patch.

gandhi56 commented:
ping @dtcxzyw

gandhi56 commented:
ping @fhahn

gandhi56 commented:
ping

nikic commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -201,61 +201,54 @@ CleanupPointerRootUsers(GlobalVariable *GV,
+  // to remove it before removing the store. Remove both the store and the
+  // computation chain if it is safe to do so.
+  while (!Writes.empty()) {
+    auto *Write = Writes.pop_back_val();
```
Can you use a simple for loop over Writes here? The pop_back_val doesn't seem necessary...

nikic commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -201,61 +201,54 @@ CleanupPointerRootUsers(GlobalVariable *GV,
+  // computation chain if it is safe to do so.
+  while (!Writes.empty()) {
+    auto *Write = Writes.pop_back_val();
+    auto *V = Write->getOperand(0);
```
This is going to be the stored value for stores, but the memory intrinsics this is going to be the destination pointer. I don't think applying the check to the destination pointer makes sense? For memset we should always allow removal, for memcpy it's tricky (what really matters is the contents of the copied memory there). But in any check there would have to be on the source pointer, not the destination pointer.

nikic commented on llvm/test/Transforms/GlobalOpt/cleanup-pointer-root-users-gep-constexpr.ll:
```diff
@@ -99,14 +99,14 @@ define i32 @load_gv_from_op_remove_store(ptr %p) local_unnamed_addr {
 ; CHECK-NEXT:    br label [[IF_END]]
 ; CHECK:       if.end:
-; CHECK-NEXT:    store ptr [[E]], ptr getelementptr inbounds ([3 x ptr], ptr @b, i64 0, i64 2), align 16
+; CHECK-NOT:     store ptr [[E]], ptr getelementptr inbounds ([3 x ptr], ptr @b, i64 0, i64 2), align 16
```
Rerun update_test_checks instead of manually editing the file.

nikic commented on llvm/test/tools/llvm-ar/extract.test:
```diff
@@ -1,5 +1,4 @@
 ## Test extract operation.
-# XFAIL: system-darwin
```
Unrelated changes?

gandhi56 commented:
All llvm tests pass according to my local testing.

nikic commented on llvm/test/Transforms/GlobalOpt/cleanup-pointer-root-users.ll:
```diff
@@ -12,7 +12,7 @@ define void @test1a() {
 define void @test1b(ptr %p) {
 ; CHECK-LABEL: @test1b(
-; CHECK-NEXT: store
+; CHECK-NOT: store
```
This change looks incorrect to me: `%p` might be a malloc-ed pointer here, which would leak after the transform.

gandhi56 replied:
So, in general, store instructions with a pointer as its value operand should not be removed?

nikic replied:
Not quite. The point of CleanupPointerRootUsers() is to remove assignments for which we know that the value is not dynamically allocated. So e.g. a store of a constant is fine, but a store of an unknown argument is not, because that argument could have come from malloc.

gandhi56 commented:
ping

gandhi56 commented:
ping

efriedma-quic commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -158,12 +158,12 @@ static bool isLeakCheckerRoot(GlobalVariable *GV) {
     if (isa<Constant>(V))
       return true;
-    if (!V->hasOneUse())
+    if (V->getNumUses() > 0)
```
hasNUsesOrMore

But also, we have RecursivelyDeleteTriviallyDeadInstructions if you just want to kill off dead computations.  The point of the old code was that it was trying to preserve stores of allocations with other non-trivial uses, which is why it was checking for exactly one use.

efriedma-quic commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -202,60 +212,51 @@ CleanupPointerRootUsers(GlobalVariable *GV,
+    do {
+      if (isAllocationFn(I, GetTLI))
+        break;
+      Instruction *J = dyn_cast<Instruction>(I->getOperand(0));
```
Can this cast actually fail?  isSafeComputationToRemove is supposed to check for a series of single-operand operations leading to an allocation.

efriedma-quic reviewed (commented):
With these changes, what's the difference between cleanupPointerRootUsers and CleanupConstantGlobalUsers?

gandhi56 commented:
> With these changes, what's the difference between cleanupPointerRootUsers and CleanupConstantGlobalUsers?

`CleanupConstantGlobalUsers` cleans up the trivial users of `GV`, whereas `cleanupPointerRootUsers` handles computation chains leading up to the pointer root `GV`. `cleanupPointerRootUsers` makes a call to `isSafeComputationToRemove` before removing the chain.

gandhi56 commented:
I might be able to merge the `cleanupPointerRootUsers` and `cleanupConstantGlobalUsers` if I perform checks properly. I will work on that and update this PR.

gandhi56 commented:
ping

efriedma-quic commented on llvm/lib/Transforms/IPO/GlobalOpt.cpp:
```diff
@@ -318,11 +240,30 @@ static bool CleanupConstantGlobalUsers(GlobalVariable *GV,
+        continue;
+      if (isa<Instruction, Operator>(Src)) {
+        EraseFromParent(MTI);
+        if (isSafeComputationToRemove(Src, GetTLI))
```
Is the isSafeComputationToRemove() call actually doing anything useful here?  Once you call EraseFromParent, the computation is dead; it doesn't really make any difference whether you clean up the dead instructions here, or let instcombine delete them later.

gandhi56 replied:
The purpose of `isSafeComputationToRemove(..)` at this point is only if we want to remove the computation chain in globalopt. You're right, we could let instcombine handle the removal of the computation chain as we should in my opinion to avoid redundancy. @nikic I would like to hear your thoughts on this.

efriedma-quic replied:
I was trying to say that you can just call RecursivelyDeleteTriviallyDeadInstructions without checking if the instruction is dead; it internally checks if the instruction is dead.

But more importantly, it means isSafeComputationToRemove isn't actually doing anything useful, so the end result is that this behaves identically to D69428.

efriedma-quic commented:
(I'm not exactly against killing off isLeakCheckerRoot, but last time we tried, it caused problems; see https://reviews.llvm.org/D69428 .)

gandhi56 commented:
> (I'm not exactly against killing off isLeakCheckerRoot, but last time we tried, it caused problems; see https://reviews.llvm.org/D69428 .)

I could try building valgrind to see if there are any failures.

gandhi56 committed:
[GlobalOpt] Eliminate redundant cleanup functions

Remove CleanupPointerRootUsers and
isLeakCheckerRoot. Replace calls to
CleanupPointerRootUsers with an extended version
of CleanupConstantGlobalUsers.

Fixes #64680.

Change-Id: I4c3fc52890eec1a3706cba9ecb916fd344672756

gandhi56 commented:
@efriedma-quic Is anyone familiar with how to run valgrind tests? Or any other comments on this PR?

efriedma-quic reviewed (commented):
I think the testcase is just running a program that stores the result of a malloc to an internal global variable under valgrind.  I don't have any specific examples handy; maybe someone else does.

gandhi56 commented:
@efriedma-quic I took a second look through all the cleanup routines. `isLeakCheckerRoot(GV)` seems necessary to me even though it's rarely applicable. So I am thinking to keep `cleanupPointerRootUsers` and improve it to cover cases as such as in the issue #64680. This will also prevent any potential failures in valgrind tests.

efriedma-quic commented:
Deleting stores where we can prove the stored value isn't a pointer to an allocation should be fine; it doesn't impact the controversial cases.  (But the current version of the patch is more aggressive, maybe by accident.)

gandhi56 commented:
> Deleting stores where we can prove the stored value isn't a pointer to an allocation should be fine; it doesn't impact the controversial cases. (But the current version of the patch is more aggressive, maybe by accident.)

Assuming GV is a pointer root which is never read from, the store instruction can be removed even if it's value operand is a pointer.

I think the logic implemented to decide whether or not source value operands should be removed was a little messy. Hence the test from #64680 was failing.

gandhi56 commented:
ping

vitalybuka commented:
> ping

Assuming you are expecting @efriedma-quic reply,  don't forget "note" from https://llvm.org/docs/CodeReview.html#acknowledge-all-reviewer-feedback

gandhi56 commented:
ping

gandhi56 commented:
ping

efriedma-quic commented:
If you patch isn't getting appropriate review, merging it is not an appropriate way to escalate.  Appropriate ways include privately emailing reviewers, or starting a thread on Discourse.

-----

I apologize for not continuing to review this in a timely fashion.  I should not have ignored the patch like this, after I started reviewing.

There are basically two ways forward for this patch: one, you preserve the "leak root" check so that allocation checking tools don't break.  Two, we explicitly decide to break any tools that depend on "leak root" checking (probably needs an RFC).  Your patch is currently doing neither one of those; it's still trying to straddle the middle in a way I can't understand.

nikic commented:
FWIW, my opinion on what to do here is https://github.com/llvm/llvm-project/issues/109718#issuecomment-2370770969.
